tree/treeplayer/src/TTreeProcessorMT.cxx

```diff
@@ -37,6 +37,92 @@ struct EntryCluster {
    Long64_t end;
 };
 
+// note that this routine assumes global entry numbers
+static bool ClustersAreSortedAndContiguous(const std::vector<std::vector<EntryCluster>> &cls)
+{
+   Long64_t last_end = 0ll;
+   for (const auto &fcl : cls) {
+      for (const auto &c : fcl) {
+         if (last_end != c.start)
+            return false;
+         last_end = c.end;
+      }
+   }
+   return true;
+}
+
+/// Take a vector of vectors of EntryClusters (a vector per file), filter the entries according to entryList, and
+/// and return a new vector of vectors of EntryClusters where cluster start/end entry numbers have been converted to
+/// TEntryList-local entry numbers.
+///
+/// This routine assumes that entry numbers in the TEntryList (and, if present, in the sub-entrylists) are in
+/// ascending order, i.e., for n > m:
+///   elist.GetEntry(n) + tree_offset_for_entry_from_elist(n) > elist.GetEntry(m) + tree_offset_for_entry_from_elist(m)
+static std::vector<std::vector<EntryCluster>>
+ConvertToElistClusters(std::vector<std::vector<EntryCluster>> &&clusters, TEntryList &entryList,
+                       const std::vector<std::string> &treeNames, const std::vector<std::string> &fileNames,
+                       const std::vector<Long64_t> &entriesPerFile)
+{
+   R__ASSERT(entryList.GetN() > 0); // wasteful to call this function if it has nothing to do
+   R__ASSERT(ClustersAreSortedAndContiguous(clusters));
+
+   const bool listHasGlobalEntryNumbers = entryList.GetLists() == nullptr;
+   const auto nFiles = clusters.size();
+
+   std::unique_ptr<TChain> chain;
+   if (!listHasGlobalEntryNumbers) {
+      // we need to build `chain` to be able to convert local entry numbers to global entry numbers in `Next`, below
+      chain.reset(new TChain());
+      for (auto i = 0u; i < nFiles; ++i)
+         chain->Add((fileNames[i] + "/" + treeNames[i]).c_str(), entriesPerFile[i]);
+   }
+
+   // Advance elEntry, then return global entry number corresponding to the elEntry-th entry in elist
+   // Return -1 if we are out of TEntryList entries.
+   auto Next = [listHasGlobalEntryNumbers, &chain](Long64_t &elEntry, TEntryList &elist) {
+      ++elEntry;
+      if (listHasGlobalEntryNumbers) {
+         return elist.Next();
+      } else {
+         int treenum = -1;
+         Long64_t globalEntry = elist.GetEntryAndTree(elEntry, treenum);
+         if (globalEntry != -1ll)
+            globalEntry += chain->GetTreeOffset()[treenum];
+         return globalEntry;
+      }
+   };
+
+   // the call to GetEntry also serves the purpose to reset TEntryList::fLastIndexQueried,
+   // so we can be sure TEntryList::Next will return the correct thing
+   Long64_t elistEntry = 0ll;
+   Long64_t entry = entryList.GetEntry(elistEntry);
+
+   std::vector<std::vector<EntryCluster>> elistClusters;
```

Comment:
Are those ending up being Cluster in the TTree sense or more of an arbitrary range of entries?

Reply:
they are clusters in the TTree sense, but filtered to only entries that are present in the TEntryList

Reply:
I see.  I missed the input argument's meaning (which is documented well .. i.e. my bad)

Reply:
but actually your comment made me realize that we are splitting up too much, will fix